src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -0,0 +1,69 @@
+package is.hail.expr.ir
+
+import is.hail.expr.{BaseIR, TBoolean, TFloat32, TFloat64, TInt32, TInt64, TStruct, TVoid, Type, TArray}
+
+object IR {
+  def seq(stmts: IR*)  = new Seq(stmts.toArray)
+}
+
+sealed trait IR extends BaseIR {
+  def typ: Type
+
+  override def children: IndexedSeq[BaseIR] = ???
+
+  override def copy(newChildren: IndexedSeq[BaseIR]): BaseIR = ???
+}
+sealed case class I32(x: Int) extends IR { val typ = TInt32 }
+sealed case class I64(x: Long) extends IR { val typ = TInt64 }
+sealed case class F32(x: Float) extends IR { val typ = TFloat32 }
+sealed case class F64(x: Double) extends IR { val typ = TFloat64 }
+sealed case class True() extends IR { val typ = TBoolean }
+sealed case class False() extends IR { val typ = TBoolean }
+
+sealed case class NA(typ: Type) extends IR
+sealed case class MapNA(name: String, value: IR, body: IR, var typ: Type = null) extends IR
+sealed case class IsNA(value: IR) extends IR { val typ = TBoolean }
+
+sealed case class If(cond: IR, cnsq: IR, altr: IR, var typ: Type = null) extends IR
+
+sealed case class Let(name: String, value: IR, body: IR, var typ: Type = null) extends IR
+sealed case class Ref(name: String, var typ: Type = null) extends IR
+
+sealed case class ApplyPrimitive(op: String, args: Array[IR], var typ: Type = null) extends IR {
```

Comment:
As per our discussion, I think primitive type operations (add, etc.) should be in the IR.  I imagine:

```
case class Binary(op: BinaryOp, larg1: IR, arg2: IR, ...) extends IR
case class Unary(op: UnaryOp, arg: IR, ...) extends IR
```

And then another tagged union for operations (neg, add, mul, etc.)